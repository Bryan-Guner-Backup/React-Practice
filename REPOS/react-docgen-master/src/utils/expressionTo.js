/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

/*eslint no-loop-func: 0, no-use-before-define: 0*/

import { namedTypes as t } from 'ast-types';
import resolveToValue from './resolveToValue';
import type { Importer } from '../types';

/**
 * Splits a MemberExpression or CallExpression into parts.
 * E.g. foo.bar.baz becomes ['foo', 'bar', 'baz']
 */
function toArray(path: NodePath, importer: Importer): Array<string> {
  const parts = [path];
  let result = [];

  while (parts.length > 0) {
    path = parts.shift();
    const node = path.node;
    if (t.CallExpression.check(node)) {
      parts.push(path.get('callee'));
      continue;
    } else if (t.MemberExpression.check(node)) {
      parts.push(path.get('object'));
      if (node.computed) {
        const resolvedPath = resolveToValue(path.get('property'), importer);
        if (resolvedPath !== undefined) {
          result = result.concat(toArray(resolvedPath, importer));
        } else {
          result.push('<computed>');
        }
      } else {
        result.push(node.property.name);
      }
      continue;
    } else if (t.Identifier.check(node)) {
      result.push(node.name);
      continue;
    } else if (t.Literal.check(node)) {
      result.push(node.raw);
      continue;
    } else if (t.ThisExpression.check(node)) {
      result.push('this');
      continue;
    } else if (t.ObjectExpression.check(node)) {
      const properties = path.get('properties').map(function (property) {
        return (
          toString(property.get('key'), importer) +
          ': ' +
          toString(property.get('value'), importer)
        );
      });
      result.push('{' + properties.join(', ') + '}');
      continue;
    } else if (t.ArrayExpression.check(node)) {
      result.push(
        '[' +
          path
            .get('elements')
            .map(function (el) {
              return toString(el, importer);
            })
            .join(', ') +
          ']',
      );
      continue;
    }
  }

  return result.reverse();
}

/**
 * Creates a string representation of a member expression.
 */
function toString(path: NodePath, importer: Importer): string {
  return toArray(path, importer).join('.');
}

export { toString as String, toArray as Array };
