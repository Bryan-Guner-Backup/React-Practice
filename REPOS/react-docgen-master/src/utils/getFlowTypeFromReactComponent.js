/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import type Documentation from '../Documentation';
import getMemberValuePath from './getMemberValuePath';
import getTypeAnnotation from './getTypeAnnotation';
import getTypeParameters, { type TypeParameters } from './getTypeParameters';
import isReactComponentClass from './isReactComponentClass';
import isReactForwardRefCall from './isReactForwardRefCall';
import resolveGenericTypeAnnotation from './resolveGenericTypeAnnotation';
import resolveToValue from './resolveToValue';
import type { Importer } from '../types';

function getStatelessPropsPath(componentDefinition, importer): NodePath {
  const value = resolveToValue(componentDefinition, importer);
  if (isReactForwardRefCall(value, importer)) {
    const inner = resolveToValue(value.get('arguments', 0), importer);
    return inner.get('params', 0);
  }
  return value.get('params', 0);
}

/**
 * Given an React component (stateless or class) tries to find the
 * flow type for the props. If not found or not one of the supported
 * component types returns null.
 */
export default (path: NodePath, importer: Importer): ?NodePath => {
  let typePath: ?NodePath = null;

  if (isReactComponentClass(path, importer)) {
    const superTypes = path.get('superTypeParameters');

    if (superTypes.value) {
      const params = superTypes.get('params');
      if (params.value.length === 3) {
        typePath = params.get(1);
      } else {
        typePath = params.get(0);
      }
    } else {
      const propsMemberPath = getMemberValuePath(path, 'props', importer);
      if (!propsMemberPath) {
        return null;
      }

      typePath = getTypeAnnotation(propsMemberPath.parentPath);
    }

    return typePath;
  }

  const propsParam = getStatelessPropsPath(path, importer);

  if (propsParam) {
    typePath = getTypeAnnotation(propsParam);
  }

  return typePath;
};

export function applyToFlowTypeProperties(
  documentation: Documentation,
  path: NodePath,
  callback: (propertyPath: NodePath, typeParams: ?TypeParameters) => void,
  typeParams?: ?TypeParameters,
  importer: Importer,
) {
  if (path.node.properties) {
    path
      .get('properties')
      .each(propertyPath => callback(propertyPath, typeParams));
  } else if (path.node.members) {
    path
      .get('members')
      .each(propertyPath => callback(propertyPath, typeParams));
  } else if (path.node.type === 'InterfaceDeclaration') {
    if (path.node.extends) {
      applyExtends(documentation, path, callback, typeParams, importer);
    }

    path
      .get('body', 'properties')
      .each(propertyPath => callback(propertyPath, typeParams));
  } else if (path.node.type === 'TSInterfaceDeclaration') {
    if (path.node.extends) {
      applyExtends(documentation, path, callback, typeParams, importer);
    }

    path
      .get('body', 'body')
      .each(propertyPath => callback(propertyPath, typeParams));
  } else if (
    path.node.type === 'IntersectionTypeAnnotation' ||
    path.node.type === 'TSIntersectionType'
  ) {
    path
      .get('types')
      .each(typesPath =>
        applyToFlowTypeProperties(
          documentation,
          typesPath,
          callback,
          typeParams,
          importer,
        ),
      );
  } else if (path.node.type !== 'UnionTypeAnnotation') {
    // The react-docgen output format does not currently allow
    // for the expression of union types
    const typePath = resolveGenericTypeAnnotation(path, importer);
    if (typePath) {
      applyToFlowTypeProperties(
        documentation,
        typePath,
        callback,
        typeParams,
        importer,
      );
    }
  }
}

function applyExtends(documentation, path, callback, typeParams, importer) {
  path.get('extends').each((extendsPath: NodePath) => {
    const resolvedPath = resolveGenericTypeAnnotation(extendsPath, importer);
    if (resolvedPath) {
      if (resolvedPath.node.typeParameters && extendsPath.node.typeParameters) {
        typeParams = getTypeParameters(
          resolvedPath.get('typeParameters'),
          extendsPath.get('typeParameters'),
          typeParams,
          importer,
        );
      }
      applyToFlowTypeProperties(
        documentation,
        resolvedPath,
        callback,
        typeParams,
        importer,
      );
    } else {
      const id =
        extendsPath.node.id ||
        extendsPath.node.typeName ||
        extendsPath.node.expression;
      if (id && id.type === 'Identifier') {
        documentation.addComposes(id.name);
      }
    }
  });
}
