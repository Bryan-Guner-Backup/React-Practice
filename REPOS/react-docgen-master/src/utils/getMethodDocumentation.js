/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import { namedTypes as t } from 'ast-types';
import { getDocblock } from './docblock';
import getFlowType from './getFlowType';
import getTSType from './getTSType';
import getParameterName from './getParameterName';
import getPropertyName from './getPropertyName';
import getTypeAnnotation from './getTypeAnnotation';
import type { FlowTypeDescriptor, Importer } from '../types';
import resolveToValue from './resolveToValue';

type MethodParameter = {
  name: string,
  type?: ?FlowTypeDescriptor,
  optional?: boolean,
};

type MethodReturn = {
  type: ?FlowTypeDescriptor,
};

type MethodDocumentation = {
  name: string,
  docblock: ?string,
  modifiers: Array<string>,
  params: Array<MethodParameter>,
  returns: ?MethodReturn,
};

function getMethodFunctionExpression(methodPath, importer) {
  const exprPath = t.AssignmentExpression.check(methodPath.node)
    ? methodPath.get('right')
    : methodPath.get('value');
  return resolveToValue(exprPath, importer);
}

function getMethodParamsDoc(methodPath, importer) {
  const params = [];
  const functionExpression = getMethodFunctionExpression(methodPath, importer);

  // Extract param flow types.
  functionExpression.get('params').each(paramPath => {
    let type = null;
    const typePath = getTypeAnnotation(paramPath);
    if (typePath && t.Flow.check(typePath.node)) {
      type = getFlowType(typePath, null, importer);
      if (t.GenericTypeAnnotation.check(typePath.node)) {
        type.alias = typePath.node.id.name;
      }
    } else if (typePath) {
      type = getTSType(typePath, null, importer);
      if (t.TSTypeReference.check(typePath.node)) {
        type.alias = typePath.node.typeName.name;
      }
    }

    const param = {
      name: getParameterName(paramPath),
      optional: paramPath.node.optional,
      type,
    };

    params.push(param);
  });

  return params;
}

// Extract flow return type.
function getMethodReturnDoc(methodPath, importer) {
  const functionExpression = getMethodFunctionExpression(methodPath, importer);

  if (functionExpression.node.returnType) {
    const returnType = getTypeAnnotation(functionExpression.get('returnType'));
    if (returnType && t.Flow.check(returnType.node)) {
      return { type: getFlowType(returnType, null, importer) };
    } else if (returnType) {
      return { type: getTSType(returnType, null, importer) };
    }
  }

  return null;
}

function getMethodModifiers(methodPath) {
  if (t.AssignmentExpression.check(methodPath.node)) {
    return ['static'];
  }

  // Otherwise this is a method/property node

  const modifiers = [];

  if (methodPath.node.static) {
    modifiers.push('static');
  }

  if (methodPath.node.kind === 'get' || methodPath.node.kind === 'set') {
    modifiers.push(methodPath.node.kind);
  }

  const functionExpression = methodPath.get('value').node;
  if (functionExpression.generator) {
    modifiers.push('generator');
  }
  if (functionExpression.async) {
    modifiers.push('async');
  }

  return modifiers;
}

function getMethodName(methodPath, importer) {
  if (
    t.AssignmentExpression.check(methodPath.node) &&
    t.MemberExpression.check(methodPath.node.left)
  ) {
    const left = methodPath.node.left;
    const property = left.property;
    if (!left.computed) {
      return property.name;
    }
    if (t.Literal.check(property)) {
      return String(property.value);
    }
    return null;
  }
  return getPropertyName(methodPath, importer);
}

function getMethodAccessibility(methodPath) {
  if (t.AssignmentExpression.check(methodPath.node)) {
    return null;
  }

  // Otherwise this is a method/property node
  return methodPath.node.accessibility;
}

function getMethodDocblock(methodPath) {
  if (t.AssignmentExpression.check(methodPath.node)) {
    let path = methodPath;
    do {
      path = path.parent;
    } while (path && !t.ExpressionStatement.check(path.node));
    if (path) {
      return getDocblock(path);
    }
    return null;
  }

  // Otherwise this is a method/property node
  return getDocblock(methodPath);
}

// Gets the documentation object for a component method.
// Component methods may be represented as class/object method/property nodes
// or as assignment expresions of the form `Component.foo = function() {}`
export default function getMethodDocumentation(
  methodPath: NodePath,
  importer: Importer,
): ?MethodDocumentation {
  if (getMethodAccessibility(methodPath) === 'private') {
    return null;
  }

  const name = getMethodName(methodPath, importer);
  if (!name) return null;

  return {
    name,
    docblock: getMethodDocblock(methodPath),
    modifiers: getMethodModifiers(methodPath),
    params: getMethodParamsDoc(methodPath, importer),
    returns: getMethodReturnDoc(methodPath, importer),
  };
}
