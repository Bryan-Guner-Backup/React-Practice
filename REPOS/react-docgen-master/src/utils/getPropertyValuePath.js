/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import { namedTypes as t } from 'ast-types';
import getPropertyName from './getPropertyName';
import type { Importer } from '../types';

/**
 * Given an ObjectExpression, this function returns the path of the value of
 * the property with name `propertyName`.
 */
export default function getPropertyValuePath(
  path: NodePath,
  propertyName: string,
  importer: Importer,
): ?NodePath {
  t.ObjectExpression.assert(path.node);

  return path
    .get('properties')
    .filter(
      propertyPath => getPropertyName(propertyPath, importer) === propertyName,
    )
    .map(propertyPath => propertyPath.get('value'))[0];
}
