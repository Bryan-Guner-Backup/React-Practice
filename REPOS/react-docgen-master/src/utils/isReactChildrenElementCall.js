/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import { namedTypes as t } from 'ast-types';
import isReactModuleName from './isReactModuleName';
import match from './match';
import resolveToModule from './resolveToModule';
import type { Importer } from '../types';

/**
 * Returns true if the expression is a function call of the form
 * `React.Children.only(...)`.
 */
export default function isReactChildrenElementCall(
  path: NodePath,
  importer: Importer,
): boolean {
  if (t.ExpressionStatement.check(path.node)) {
    path = path.get('expression');
  }

  if (!match(path.node, { callee: { property: { name: 'only' } } })) {
    return false;
  }

  const calleeObj = path.get('callee', 'object');
  const module = resolveToModule(calleeObj, importer);

  if (!match(calleeObj, { value: { property: { name: 'Children' } } })) {
    return false;
  }

  return Boolean(module && isReactModuleName(module));
}
