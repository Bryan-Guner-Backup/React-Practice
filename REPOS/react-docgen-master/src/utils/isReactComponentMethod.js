/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import { namedTypes as t } from 'ast-types';
import getPropertyName from './getPropertyName';
import type { Importer } from '../types';

const componentMethods = [
  'componentDidMount',
  'componentDidReceiveProps',
  'componentDidUpdate',
  'componentWillMount',
  'UNSAFE_componentWillMount',
  'componentWillReceiveProps',
  'UNSAFE_componentWillReceiveProps',
  'componentWillUnmount',
  'componentWillUpdate',
  'UNSAFE_componentWillUpdate',
  'getChildContext',
  'getDefaultProps',
  'getInitialState',
  'render',
  'shouldComponentUpdate',
  'getDerivedStateFromProps',
  'getDerivedStateFromError',
  'getSnapshotBeforeUpdate',
  'componentDidCatch',
];

/**
 * Returns if the method path is a Component method.
 */
export default function (methodPath: NodePath, importer: Importer): boolean {
  if (
    !t.MethodDefinition.check(methodPath.node) &&
    !t.Property.check(methodPath.node)
  ) {
    return false;
  }

  const name = getPropertyName(methodPath, importer);
  return !!name && componentMethods.indexOf(name) !== -1;
}
