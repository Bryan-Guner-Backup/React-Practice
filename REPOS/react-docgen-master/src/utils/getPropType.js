/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

/*eslint no-use-before-define: 0*/

import { namedTypes as t } from 'ast-types';
import { getDocblock } from '../utils/docblock';
import getMembers from './getMembers';
import getPropertyName from './getPropertyName';
import isRequiredPropType from '../utils/isRequiredPropType';
import printValue from './printValue';
import resolveToValue from './resolveToValue';
import resolveObjectKeysToArray from './resolveObjectKeysToArray';
import resolveObjectValuesToArray from './resolveObjectValuesToArray';
import type { PropTypeDescriptor, PropDescriptor, Importer } from '../types';

function getEnumValues(path: NodePath, importer: Importer) {
  const values = [];

  path.get('elements').each(function (elementPath) {
    if (t.SpreadElement.check(elementPath.node)) {
      const value = resolveToValue(elementPath.get('argument'), importer);

      if (t.ArrayExpression.check(value.node)) {
        // if the SpreadElement resolved to an Array, add all their elements too
        return values.push(...getEnumValues(value, importer));
      } else {
        // otherwise we'll just print the SpreadElement itself
        return values.push({
          value: printValue(elementPath),
          computed: !t.Literal.check(elementPath.node),
        });
      }
    }

    // try to resolve the array element to it's value
    const value = resolveToValue(elementPath, importer);
    return values.push({
      value: printValue(value),
      computed: !t.Literal.check(value.node),
    });
  });

  return values;
}

function getPropTypeOneOf(argumentPath: NodePath, importer: Importer) {
  const type: PropTypeDescriptor = { name: 'enum' };
  let value = resolveToValue(argumentPath, importer);
  if (!t.ArrayExpression.check(value.node)) {
    value =
      resolveObjectKeysToArray(value, importer) ||
      resolveObjectValuesToArray(value, importer);
    if (value) {
      type.value = getEnumValues(value, importer);
    } else {
      // could not easily resolve to an Array, let's print the original value
      type.computed = true;
      type.value = printValue(argumentPath);
    }
  } else {
    type.value = getEnumValues(value, importer);
  }
  return type;
}

function getPropTypeOneOfType(argumentPath: NodePath, importer: Importer) {
  const type: PropTypeDescriptor = { name: 'union' };
  if (!t.ArrayExpression.check(argumentPath.node)) {
    type.computed = true;
    type.value = printValue(argumentPath);
  } else {
    type.value = argumentPath.get('elements').map(function (itemPath) {
      const descriptor: PropTypeDescriptor = getPropType(itemPath, importer);
      const docs = getDocblock(itemPath);
      if (docs) {
        descriptor.description = docs;
      }
      return descriptor;
    });
  }
  return type;
}

function getPropTypeArrayOf(argumentPath: NodePath, importer: Importer) {
  const type: PropTypeDescriptor = { name: 'arrayOf' };

  const docs = getDocblock(argumentPath);
  if (docs) {
    type.description = docs;
  }

  const subType = getPropType(argumentPath, importer);

  if (subType.name === 'unknown') {
    type.value = printValue(argumentPath);
    type.computed = true;
  } else {
    type.value = subType;
  }
  return type;
}

function getPropTypeObjectOf(argumentPath: NodePath, importer: Importer) {
  const type: PropTypeDescriptor = { name: 'objectOf' };

  const docs = getDocblock(argumentPath);
  if (docs) {
    type.description = docs;
  }

  const subType = getPropType(argumentPath, importer);

  if (subType.name === 'unknown') {
    type.value = printValue(argumentPath);
    type.computed = true;
  } else {
    type.value = subType;
  }
  return type;
}

/**
 * Handles shape and exact prop types
 */
function getPropTypeShapish(
  name: 'shape' | 'exact',
  argumentPath: NodePath,
  importer: Importer,
) {
  const type: PropTypeDescriptor = { name };
  if (!t.ObjectExpression.check(argumentPath.node)) {
    argumentPath = resolveToValue(argumentPath, importer);
  }

  if (t.ObjectExpression.check(argumentPath.node)) {
    const value = {};
    argumentPath.get('properties').each(function (propertyPath) {
      if (propertyPath.get('type').value === t.SpreadElement.name) {
        // It is impossible to resolve a name for a spread element
        return;
      }

      const propertyName = getPropertyName(propertyPath, importer);
      if (!propertyName) return;
      const descriptor: PropDescriptor | PropTypeDescriptor = getPropType(
        propertyPath.get('value'),
        importer,
      );
      const docs = getDocblock(propertyPath);
      if (docs) {
        descriptor.description = docs;
      }
      descriptor.required = isRequiredPropType(propertyPath.get('value'));
      value[propertyName] = descriptor;
    });
    type.value = value;
  }

  if (!type.value) {
    type.value = printValue(argumentPath);
    type.computed = true;
  }

  return type;
}

// eslint-disable-next-line no-unused-vars
function getPropTypeInstanceOf(argumentPath: NodePath, importer: Importer) {
  return {
    name: 'instanceOf',
    value: printValue(argumentPath),
  };
}

const simplePropTypes = [
  'array',
  'bool',
  'func',
  'number',
  'object',
  'string',
  'any',
  'element',
  'node',
  'symbol',
  'elementType',
];

const propTypes = new Map([
  ['oneOf', getPropTypeOneOf],
  ['oneOfType', getPropTypeOneOfType],
  ['instanceOf', getPropTypeInstanceOf],
  ['arrayOf', getPropTypeArrayOf],
  ['objectOf', getPropTypeObjectOf],
  ['shape', getPropTypeShapish.bind(null, 'shape')],
  ['exact', getPropTypeShapish.bind(null, 'exact')],
]);

/**
 * Tries to identify the prop type by inspecting the path for known
 * prop type names. This method doesn't check whether the found type is actually
 * from React.PropTypes. It simply assumes that a match has the same meaning
 * as the React.PropTypes one.
 *
 * If there is no match, "custom" is returned.
 */
export default function getPropType(
  path: NodePath,
  importer: Importer,
): PropTypeDescriptor {
  let descriptor;
  getMembers(path, true).some(member => {
    const node = member.path.node;
    let name;
    if (t.Literal.check(node)) {
      name = node.value;
    } else if (t.Identifier.check(node) && !member.computed) {
      name = node.name;
    }
    if (name) {
      if (simplePropTypes.includes(name)) {
        descriptor = { name };
        return true;
      } else if (propTypes.has(name) && member.argumentsPath) {
        // $FlowIssue
        descriptor = propTypes.get(name)(member.argumentsPath.get(0), importer);
        return true;
      }
    }
  });
  if (!descriptor) {
    const node = path.node;
    if (t.Identifier.check(node) && simplePropTypes.includes(node.name)) {
      descriptor = { name: node.name };
    } else if (
      t.CallExpression.check(node) &&
      t.Identifier.check(node.callee) &&
      propTypes.has(node.callee.name)
    ) {
      // $FlowIssue
      descriptor = propTypes.get(node.callee.name)(path.get('arguments', 0));
    } else {
      descriptor = { name: 'custom', raw: printValue(path) };
    }
  }
  // $FlowIssue
  return descriptor;
}
