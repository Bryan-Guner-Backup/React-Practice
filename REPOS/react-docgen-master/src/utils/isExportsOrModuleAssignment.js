/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import { namedTypes as t } from 'ast-types';
import * as expressionTo from './expressionTo';
import type { Importer } from '../types';

/**
 * Returns true if the expression is of form `exports.foo = ...;` or
 * `modules.exports = ...;`.
 */
export default function isExportsOrModuleAssignment(
  path: NodePath,
  importer: Importer,
): boolean {
  if (t.ExpressionStatement.check(path.node)) {
    path = path.get('expression');
  }
  if (
    !t.AssignmentExpression.check(path.node) ||
    !t.MemberExpression.check(path.node.left)
  ) {
    return false;
  }

  const exprArr = expressionTo.Array(path.get('left'), importer);
  return (
    (exprArr[0] === 'module' && exprArr[1] === 'exports') ||
    exprArr[0] === 'exports'
  );
}
