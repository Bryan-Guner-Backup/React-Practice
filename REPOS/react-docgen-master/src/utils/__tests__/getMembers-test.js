/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { expression } from '../../../tests/utils';
import getMembers from '../getMembers';

describe('getMembers', () => {
  it('finds all "members" "inside" a MemberExpression', () => {
    const members = getMembers(expression('foo.bar(123)(456)[baz][42]'));

    expect(members).toMatchSnapshot();
  });
});
