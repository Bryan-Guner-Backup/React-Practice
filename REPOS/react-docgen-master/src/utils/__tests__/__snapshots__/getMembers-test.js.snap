// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMembers finds all "members" "inside" a MemberExpression 1`] = `
Array [
  Object {
    "argumentsPath": Node {
      "arguments": Array [
        Node {
          "end": 12,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 12,
              "line": 1,
            },
            "filename": undefined,
            "identifierName": undefined,
            "start": Position {
              "column": 9,
              "line": 1,
            },
          },
          "range": undefined,
          "raw": "123",
          "start": 9,
          "trailingComments": undefined,
          "type": "Literal",
          "value": 123,
        },
      ],
      "callee": Node {
        "computed": false,
        "end": 8,
        "extra": undefined,
        "innerComments": undefined,
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 8,
            "line": 1,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 1,
            "line": 1,
          },
        },
        "object": Node {
          "end": 4,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 4,
              "line": 1,
            },
            "filename": undefined,
            "identifierName": "foo",
            "start": Position {
              "column": 1,
              "line": 1,
            },
          },
          "name": "foo",
          "range": undefined,
          "start": 1,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "optional": false,
        "property": Node {
          "end": 8,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 8,
              "line": 1,
            },
            "filename": undefined,
            "identifierName": "bar",
            "start": Position {
              "column": 5,
              "line": 1,
            },
          },
          "name": "bar",
          "range": undefined,
          "start": 5,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "range": undefined,
        "start": 1,
        "trailingComments": undefined,
        "type": "MemberExpression",
      },
      "end": 13,
      "extra": undefined,
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 13,
          "line": 1,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 1,
          "line": 1,
        },
      },
      "optional": false,
      "range": undefined,
      "start": 1,
      "trailingComments": undefined,
      "type": "CallExpression",
    },
    "computed": false,
    "path": Node {
      "end": 8,
      "extra": undefined,
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 8,
          "line": 1,
        },
        "filename": undefined,
        "identifierName": "bar",
        "start": Position {
          "column": 5,
          "line": 1,
        },
      },
      "name": "bar",
      "range": undefined,
      "start": 5,
      "trailingComments": undefined,
      "type": "Identifier",
    },
  },
  Object {
    "argumentsPath": null,
    "computed": true,
    "path": Node {
      "end": 22,
      "extra": undefined,
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 22,
          "line": 1,
        },
        "filename": undefined,
        "identifierName": "baz",
        "start": Position {
          "column": 19,
          "line": 1,
        },
      },
      "name": "baz",
      "range": undefined,
      "start": 19,
      "trailingComments": undefined,
      "type": "Identifier",
    },
  },
  Object {
    "argumentsPath": null,
    "computed": true,
    "path": Node {
      "end": 26,
      "innerComments": undefined,
      "leadingComments": undefined,
      "loc": SourceLocation {
        "end": Position {
          "column": 26,
          "line": 1,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 24,
          "line": 1,
        },
      },
      "range": undefined,
      "raw": "42",
      "start": 24,
      "trailingComments": undefined,
      "type": "Literal",
      "value": 42,
    },
  },
]
`;
